JetUtilities/src/L2Creator.cc

```diff
@@ -431,372 +550,374 @@ bool L2Creator::getL3Rsp() {
+  //
+  // set the best parameters and chi2 to the fit function
+  //
+  TF1 * ffh = gabscor->GetFunction("fit");
+  for (unsigned int np=0;np < bestPars.size() ; np++){
+    ffh->SetParameter(np,bestPars[np]);
+    fabscor->SetParameter(np,bestPars[np]);
+  }
+  fabscor->SetChisquare(bestRChi2 * fabscor->GetNDF());
+  ffh->SetChisquare(bestRChi2 * fabscor->GetNDF());
+
+  //
+  // warn if the fit diverges at low pt
+  //
+  if (fabscor->Integral(0,10) > 60)
+    cout << "\t***ERROR***, fit for histo " << gabscor->GetName() << " diverges at low pt" << endl;
+
+  //   
+  // check for failed fits
+  // a chi2 of zero is symptomatic of a failed fit.
+  //
+  if (bestRChi2 < 0.001) {
+    cout<<"\t***ERROR***, FIT HAS FAILED for histo "<<gabscor->GetName()
+      <<" which has a reduced chi2="<<bestRChi2
+      <<" after "<<fitIter<<" iterations. "<<endl;
+  }
+
+  //
+  // check for large reduced chi2's
+  // above 10 is a plain error; between 5 and 10 is a warning
+  //
+  if (bestRChi2 > 5){
+    if (bestRChi2 > 10)
+      cout<<"\t***ERROR***,";
+    else
+      cout<<"\tWARNING,";
+
+    cout<<" fit for histo "<<gabscor->GetName()
+      <<" has a reduced chi2="<<bestRChi2
+      <<" after "<<fitIter<<" iterations"<<endl;
+  }
 }
 
 //______________________________________________________________________________
 void L2Creator::writeTextFileForCurrentAlgorithm() {
-	TString txtfilename = outputDir + era + "_L2Relative_" + ji->getAlias() + ".txt";
-	ofstream fout(txtfilename);
-	fout.setf(ios::right); 
-
-	unsigned int vector_size = 0;
-    if(l2l3) vector_size = vabscor_eta.size(); //For L2L3 Corrections Together
-    else vector_size = vrelcor_eta.size(); //For L2 & L3 Corrections Separate
-    for (unsigned int ieta=0;ieta<vector_size;ieta++) { 
-    	TGraph* grelcor;
-      	if(l2l3) grelcor = vabscor_eta[ieta]; //For L2L3 Corrections Together
-      	else grelcor = vrelcor_eta[ieta]; //For L2 & L3 Corrections Separate
-      	TF1* frelcor = (TF1*)grelcor->GetListOfFunctions()->Last();
-      	if(frelcor!=0) {
-      		if(ieta==0 || (ieta==1 && delphes))
-               fout<<"{1 JetEta 1 JetPt max(0.0001,"<<frelcor->GetTitle()<<") Correction L2Relative}"<<endl;
-               //fout<<"{1 JetEta 1 JetPt max(0.0001,"<<frelcor->GetExpFormula()<<") Correction L2Relative}"<<endl;
-      		double  etamin  = hl_jetpt.minimum(0,ieta);
-      		double  etamax  = hl_jetpt.maximum(0,ieta);
-      		double  ptmin = grelcor->GetX()[0];
-      		double  ptmax = grelcor->GetX()[grelcor->GetN()-1];
-      		fout<<setw(11)<<etamin
-      		<<setw(11)<<etamax
-        	    <<setw(11)<<(int)(frelcor->GetNpar()+2) //Number of parameters + 2 
-        	    <<setw(12)<<ptmin
-        	    <<setw(12)<<ptmax;
-        	    for(int p=0; p<frelcor->GetNpar(); p++) {
-        	    	fout<<setw(13)<<frelcor->GetParameter(p);
-        	    }
-        	    fout<<endl;
-        }
+  TString txtfilename = outputDir + "/" + era + "_L3Absolute_" + ji->getAlias() + ".txt";
```

Comment:
Hi Alexx,

I changed the name because it has issue with the tau processor, I put 
below [1] the email from my college.

After that Christian (I added in cc) voted for ..._L3Absolute_...

So we could put a flag or change the tau processor...

Cheers,
Betty


[1]
###############

Hi Christian and Betty,

I have the code mostly working and one remaining question.

The question is about which name/tag the corrections are supposed to have. Betty called the files “[..]L2Relative[..]”, so this is what I used.

This does however have the effect that the pat:::Taus are not corrected (*) (the corrections are still saved and accessible though). The solution is to either call the files - and thereby the corrections - “L2L3Residual” or “L3Absolute”, or to modify the PATTauProducer code to also accept L2Relative or whatever else we think it should be called.

Let me know once you have the updated corrections so I can run the tuple production.

Cheers,
Jan

(*)
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/PhysicsTools/PatAlgos/plugins/PATTauProducer.cc#L286

(**) Note that there is no L2L3 without “Residual” (with, at least for me, “Residual” implying data) herehttps://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/CondFormats/JetMETObjects/interface/JetCorrectorParameters.h#L116


###############


Le 22/04/2016 16:45, Alexx Perloff a écrit :
>
> In JetUtilities/src/L2Creator.cc 
> <https://github.com/cms-jet/JetMETAnalysis/pull/10#discussion_r60740520>:
>
> > -               //fout<<"{1 JetEta 1 JetPt max(0.0001,"<<frelcor->GetExpFormula()<<") Correction L2Relative}"<<endl;
> > -      		double  etamin  = hl_jetpt.minimum(0,ieta);
> > -      		double  etamax  = hl_jetpt.maximum(0,ieta);
> > -      		double  ptmin = grelcor->GetX()[0];
> > -      		double  ptmax = grelcor->GetX()[grelcor->GetN()-1];
> > -      		fout<<setw(11)<<etamin
> > -      		<<setw(11)<<etamax
> > -        	    <<setw(11)<<(int)(frelcor->GetNpar()+2) //Number of parameters + 2
> > -        	    <<setw(12)<<ptmin
> > -        	    <<setw(12)<<ptmax;
> > -        	    for(int p=0; p<frelcor->GetNpar(); p++) {
> > -        	    	fout<<setw(13)<<frelcor->GetParameter(p);
> > -        	    }
> > -        	    fout<<endl;
> > -        }
> > +  TString txtfilename = outputDir + "/" + era + "_L3Absolute_" + ji->getAlias() + ".txt";
>
> So it was just a choice? If so, why change anything at all? 
> Nevertheless, I'm going to insist that you put this line back the way 
> it was (i.e. use L2Relative). This code was meant to make L2 
> corrections and does not make L3 corrections, that's a difference 
> process/code all together. Also this would "break" the current JEC 
> workflow (we would need to rename each file that came out).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub 
> <https://github.com/cms-jet/JetMETAnalysis/pull/10/files/0f73e5f4798a6aa76f70304590ae8f2aafd9668d#r60740520>
>